modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/IngestFailureStoreMetricsIT.java

```diff
@@ -0,0 +1,359 @@
+        logger.info(
+            "--> running [{}] with data stream [{}], template [{}] and pipeline [{}]",
+            getTestName(),
+            dataStream,
+            template,
+            pipeline
+        );
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return List.of(DataStreamsPlugin.class, CustomIngestTestPlugin.class, TestTelemetryPlugin.class, MapperExtrasPlugin.class);
+    }
+
+    public void testNoPipelineNoFailures() throws IOException {
+        putComposableIndexTemplate(true);
+        createDataStream();
+
+        int nrOfDocs = randomIntBetween(5, 10);
+        indexDocs(dataStream, nrOfDocs, null);
+
+        var measurements = collectTelemetry();
+        assertMeasurements(measurements.get(FailureStoreMetrics.METRIC_TOTAL), nrOfDocs, dataStream);
+        assertEquals(0, measurements.get(FailureStoreMetrics.METRIC_FAILURE_STORE).size());
+        assertEquals(0, measurements.get(FailureStoreMetrics.METRIC_REJECTED).size());
+    }
+
+    public void testFailingPipelineNoFailureStore() throws IOException {
+        putComposableIndexTemplate(false);
+        createDataStream();
+        createBasicPipeline("fail");
+
+        int nrOfSuccessfulDocs = randomIntBetween(5, 10);
+        indexDocs(dataStream, nrOfSuccessfulDocs, null);
+        int nrOfFailingDocs = randomIntBetween(5, 10);
+        indexDocs(dataStream, nrOfFailingDocs, pipeline);
+
+        var measurements = collectTelemetry();
+        assertMeasurements(measurements.get(FailureStoreMetrics.METRIC_TOTAL), nrOfSuccessfulDocs + nrOfFailingDocs, dataStream);
+        assertEquals(0, measurements.get(FailureStoreMetrics.METRIC_FAILURE_STORE).size());
+        assertMeasurements(
+            measurements.get(FailureStoreMetrics.METRIC_REJECTED),
+            nrOfFailingDocs,
+            dataStream,
+            FailureStoreMetrics.ErrorLocation.PIPELINE,
+            false
+        );
+    }
+
+    public void testFailingPipelineWithFailureStore() throws IOException {
+        putComposableIndexTemplate(true);
+        createDataStream();
+        createBasicPipeline("fail");
+
+        int nrOfSuccessfulDocs = randomIntBetween(5, 10);
+        indexDocs(dataStream, nrOfSuccessfulDocs, null);
+        int nrOfFailingDocs = randomIntBetween(5, 10);
+        indexDocs(dataStream, nrOfFailingDocs, pipeline);
+
+        var measurements = collectTelemetry();
+        assertMeasurements(measurements.get(FailureStoreMetrics.METRIC_TOTAL), nrOfSuccessfulDocs + nrOfFailingDocs, dataStream);
+        assertMeasurements(
+            measurements.get(FailureStoreMetrics.METRIC_FAILURE_STORE),
+            nrOfFailingDocs,
+            dataStream,
+            FailureStoreMetrics.ErrorLocation.PIPELINE
+        );
+        assertEquals(0, measurements.get(FailureStoreMetrics.METRIC_REJECTED).size());
+    }
+
+    public void testShardFailureNoFailureStore() throws IOException {
```

Comment:
I wasn't able to come up with a way to make a redirect to the failure store fail on shard-level (i.e. get a rejection from the failure store). If anyone has any tips on that, feel free to share :)

Reply:
Perfect suggestion! Works like a charm :smile: